Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift

```diff
@@ -449,6 +449,21 @@ extension PoolStateMachine {
             return (index, context)
         }
 
+        /// Returns `true` if the connection should be explicitly closed, `false` if nothing needs to be done.
+        @inlinable
+        mutating func keepAliveFailed(_ connectionID: Connection.ID) -> Bool {
+            // We don't have to inform the ConnectionStates about any of this, because the connection will close
+            // immediately after this or is closed already
+            switch self.connections.first(where: { $0.id == connectionID })?.state {
```

Comment:
please don‘t access the connection‘s state here. the state is only visible here, to allow inlining. Instead please forward a call keepAliveFailed into the ConnectionStateMachine. Also make sure to change the state to closing, if we need to explicitly close the connection.

Reply:
The connection state is receiving the failure now and changes the state accordingly.